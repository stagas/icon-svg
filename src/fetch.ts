import { memoize } from 'memoize-pure'
import { getIconUrl, IconSets, IconSetsNames, IconsWithTypes, Icons, IconTypes, IconKinds } from '.'

const fetchMemoized = memoize(async (url: string) => {
  const response = await fetch(url)
  const text = await response.text()
  if (response.ok) return text
  else {
    console.error(url + ': ' + text)
    return `<span title="Error loading icon: ${url}">⚠️</span>`
  }
})

export interface FetchProps {
  raw?: boolean
  strokeWidth?: string | number
}

/**
 * Fetches an SVG icon.
 *
 * The network request is memoized so when this is called multiple times
 * it will only access the network the first time (or the browser cache if there already).
 *
 * It also attempts to normalize the SVG to make all the icons consistent and
 * usable out of the box without any adjustments, i.e it normalizes various values.
 *
 * Use CSS variables `--stroke` `--fill` and `--fill-secondary` to manage
 * the colors if necessary.
 *
 * When used with `strokeWidth` you can achieve fine-control on how it's being displayed.
 *
 * Use the `raw` option to skip any post processing if you don't need it and simply get
 * the raw SVG data.
 *
 * ```ts
 * svg = await fetchIconSvg({ set: 'bytesize', icon: 'heart', strokeWidth: '2px' })
 * svg = await fetchIconSvg({ set: 'flags', icon: 'gr', kind: '4x3', raw: true })
 * svg = await fetchIconSvg({
 *   set: 'boxicons',
 *   icon: 'alarm',
 *   type: 'solid',
 *   strokeWidth: '' // passing `strokeWidth=''` allows it to be managed in CSS
 * })
 * ```
 *
 * @param props
 * @param props.set Icon set
 * @param props.icon Icon name
 * @param props.type Icon type - Most of the time an entirely different set of icons
 * @param props.kind Icon kind - These are variations of the same icon
 * @param props.raw Get the raw SVG without any post processing
 * @param props.strokeWidth Normalize `stroke-width` to this value.
 *  This also applies `vector-effect="non-scaling-stroke"` which will make the
 *  `stroke-width` absolute and not relative to the size of the icon
 *  Use this option for more control over the icon's appearance.
 *  Because every dimension of the icon will need a different `stroke-width`
 *  this can be set to `''` (empty string) which then allows it to be handled with CSS.
 *
 * @returns The SVG data as a string
 */
export const fetchIconSvg = async <P extends keyof IconSets<IconSetsNames>, T extends IconTypes[P & keyof IconTypes]>(
  props: FetchProps & {
    set: P
  } & (P extends keyof IconKinds
      ? {
          icon: Icons[P]
          kind: IconKinds[P & keyof IconKinds]
        }
      : P extends keyof IconTypes
      ? {
          icon: IconsWithTypes[P & keyof IconsWithTypes][T & keyof IconsWithTypes[P & keyof IconsWithTypes]]
          type: T
        }
      : {
          icon: Icons[P & keyof Icons]
        })
) => {
  const url = getIconUrl(props as never)
  let svg = (await fetchMemoized(url)).trim()

  if (props.raw || ['emojicc', 'flags'].includes(props.set)) return svg

  svg = svg
    .replace(/(?<=^<svg[^>]*)\s(width|height)="[\d.]+"(?=.*>)/gis, '')
    .replace(/(^<svg[^>]*?)>/gis, '$1 fill="var(--stroke)">')

    // these two are weird unnecessary squares in iconpark's icons
    .replace('<rect width="48" height="48" fill="white" fill-opacity="0.01"/>', '')
    .replace('<path d="M48 0H0V48H48V0Z" fill="white" fill-opacity="0.01"/>', '')

  // normalize stroke-width
  if (props.strokeWidth != null)
    svg = svg
      .replace(/(stroke-width=")([^"]*?)(")/gi, `$1${props.strokeWidth as string}$3`)
      .replace(/(<(?:path|circle|rect|polyline)[^>]*?)(\/?>)/gis, '$1 vector-effect="non-scaling-stroke"$2')

  svg = svg
    // these are iconpark's colors but could be extended further
    .replace(/"#2F88FF"/gis, '"var(--fill)"')
    .replace(/"#43CCF8"/gis, '"var(--fill-secondary)"')

    // inherit so we can style with css
    .replace(/"(currentColor|white|black)"/gis, '"var(--stroke)"')

  return svg
}
