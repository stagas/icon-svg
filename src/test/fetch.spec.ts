import 'construct-style-sheets-polyfill'
import { inspectWithPreamble } from '@n1kk/intspector'
import { fetchIconSvg } from '../fetch'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
declare const window: any

const typeTest = (fn: string) => {
  try {
    inspectWithPreamble(`
      import { fetchIconSvg } from '../fetch'
      ;(window as any).fetch = async () => ({
        ok: true,
        text: async () => 'cool'
      })
      const result = ${fn}
    `)({ result: 'typeof result' })
    return true
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } catch (error: any) {
    return error[0].messageText.next[0].messageText
  }
}

describe('fetchIconSvg', () => {
  it('fetches icon', async () => {
    window.fetch = async () => ({
      ok: true,
      text: async () => `<svg><path stroke-width="1.33" /></svg>`,
    })
    const svg = await fetchIconSvg({ set: 'boxicons', icon: 'alarm', type: 'regular' })
    expect(svg).toMatchSnapshot()
  })

  it('raw', async () => {
    window.fetch = async () => ({
      ok: true,
      text: async () => `<svg><path stroke-width="1.33" /></svg>`,
    })
    const svg = await fetchIconSvg({ raw: true, set: 'boxicons', icon: 'alarm', type: 'regular' })
    expect(svg).toMatchSnapshot()
  })

  it('error', async () => {
    const consoleError = console.error
    let errorMessage
    console.error = (s: string) => {
      errorMessage = s
    }
    window.fetch = async () => ({
      ok: false,
      text: async () => `some error`,
    })
    const svg = await fetchIconSvg({ set: 'boxicons', icon: 'airbnb', type: 'logos' })
    expect(svg).toMatchSnapshot()
    expect(errorMessage).toMatchSnapshot()
    console.error = consoleError
  })

  it('memoizes fetch', async () => {
    window.fetch = async () => ({
      ok: true,
      text: async () => `<svg><path stroke-width="1.33" /></svg>`,
    })
    let svg
    svg = await fetchIconSvg({ set: 'boxicons', icon: 'alarm', type: 'regular' })
    expect(svg).toMatchSnapshot()
    window.fetch = async () => ({
      ok: true,
      text: async () => `something else`,
    })
    svg = await fetchIconSvg({ set: 'boxicons', icon: 'alarm', type: 'regular' })
    expect(svg).toMatchSnapshot()
    svg = await fetchIconSvg({ set: 'boxicons', icon: 'alarm', type: 'solid' })
    expect(svg).toMatchSnapshot()
  })

  it('applies strokeWidth', async () => {
    window.fetch = async () => ({
      ok: true,
      text: async () => `<svg><path stroke-width="1.33" /></svg>`,
    })
    const svg = await fetchIconSvg({ strokeWidth: '2px', set: 'boxicons', icon: 'alarm', type: 'regular' })
    expect(svg).toMatchSnapshot()
  })

  it('missing type when required', () => {
    expect(typeTest(`fetchIconSvg({ set: 'boxicons', type: 'logos', icon: '99designs' })`)).toEqual(true)
    expect(typeTest(`fetchIconSvg({ set: 'boxicons', icon: '99designs' })`)).toContain("'type' is missing")
  })

  it('missing kind when required', () => {
    expect(typeTest(`fetchIconSvg({ set: 'flags', icon: 'gr', kind: '4x3' })`)).toEqual(true)
    expect(typeTest(`fetchIconSvg({ set: 'flags', icon: 'gr' })`)).toContain("'kind' is missing")
  })
})
