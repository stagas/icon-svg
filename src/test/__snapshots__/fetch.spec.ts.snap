// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchIconSvg applies strokeWidth 1`] = `"<svg fill=\\"var(--stroke)\\"><path stroke-width=\\"2px\\"  vector-effect=\\"non-scaling-stroke\\"/></svg>"`;

exports[`fetchIconSvg error 1`] = `"<span title=\\"Error loading icon: https://cdn.jsdelivr.net/npm/boxicons@2/svg/logos/bxl-airbnb.svg\\">⚠️</span>"`;

exports[`fetchIconSvg error 2`] = `"https://cdn.jsdelivr.net/npm/boxicons@2/svg/logos/bxl-airbnb.svg: some error"`;

exports[`fetchIconSvg fetches icon 1`] = `"<svg fill=\\"var(--stroke)\\"><path stroke-width=\\"1.33\\" /></svg>"`;

exports[`fetchIconSvg memoizes fetch 1`] = `"<svg fill=\\"var(--stroke)\\"><path stroke-width=\\"1.33\\" /></svg>"`;

exports[`fetchIconSvg memoizes fetch 2`] = `"<svg fill=\\"var(--stroke)\\"><path stroke-width=\\"1.33\\" /></svg>"`;

exports[`fetchIconSvg memoizes fetch 3`] = `"something else"`;

exports[`fetchIconSvg raw 1`] = `"<svg><path stroke-width=\\"1.33\\" /></svg>"`;
